### PR TITLE
Near global one degree setup

### PR DESCRIPTION
**Pull request by glwagner**
This PR adds a near-global one degree setup to `validation/near_global_lat_lon`. This is mostly @sandreza and @simone-silvestri 's work. The setup spans -75S to 75N and is implemented on a regular `LatitudeLongitudeGrid` with 1 degree horizontal resolution and 50 vertical levels.

It's WIP right now partly because @sandreza is still sorting out some difficulties with including a `IsopycnalSkewSymmetricDiffusivity` closure to this setup. In addition, I'm going to update the `closure_tuples.jl` implementation because this setup may require a tuple of 5 closures.

As a teaser, the file `validation/near_global_lat_lon/one_degree_setups/plot_bathymetry.jl` produces

image

The bathymetry file is small (not much larger than few dozen long text files), so I've uploaded it directly into the repo from this PR. If we don't want to do this, I can reopen this PR with the same `.jl` files and download the bathymetry from `OceananigansArtifacts.jl` instead.

I'm thinking that we should add at least two one degree setups in this PR:

1. A "realistic" setup with initial conditions and surface relaxation or fluxes loaded from file (which @sandreza is working on)

2. An "idealized" setup that uses idealized and constant momentum and temperature forcing (perhaps with `constant_salinity=35.0`)

and possibly others.

Based on this experience, we can go in a similar direction with the 1/4 degree setup.

### CODE REVIEW

**Comment by glwagner** [comment]
@sandreza compiler hint for tuples of 5 closures: https://github.com/CliMA/Oceananigans.jl/pull/2463/commits/de2db1085b8ded2cd6a0eb254a66a4d1dea99daf

File: src/TurbulenceClosures/closure_tuples.jl [hunk 1]
```diff
@@ -41,6 +41,13 @@ for (f, alt_f) in zip(funcs, alt_funcs)
                   + $alt_f(i, j, k, grid, closures[3], Ks[3], args...) 
                   + $alt_f(i, j, k, grid, closures[4], Ks[4], args...))
 
+    @inline $f(i, j, k, grid, closures::Tuple{<:Any, <:Any, <:Any, <:Any, <:Any}, Ks, args...) = (
+                    $alt_f(i, j, k, grid, closures[1], Ks[1], args...)
+                  + $alt_f(i, j, k, grid, closures[2], Ks[2], args...) 
+                  + $alt_f(i, j, k, grid, closures[3], Ks[3], args...) 
+                  + $alt_f(i, j, k, grid, closures[4], Ks[4], args...)
+                  + $alt_f(i, j, k, grid, closures[5], Ks[5], args...))
```
Comment (glwagner) [inline src/TurbulenceClosures/closure_tuples.jl:49]: https://www.youtube.com/watch?v=Vsas7JI4quw

**Comment by glwagner** [comment]
Here's another teaser via @sandreza :

image

however we are still working on getting the initial conditions and forcing online (and the script that runs it) to produce this results --- hopefully more to come.

**Comment by glwagner** [comment]
Note on compilation and runtime: on my laptop, `time_step(simulation)` takes about 10 minutes to compile. After that it takes about 1s per time-step. With `dt=20minutes`, we need 26280 time-steps to reach 1 year, which would therefore take 7.3 hours on my laptop using 1 CPU.

**Comment by glwagner** [comment]
I added dropbox links to download the boundary conditions. This is what the annual cycle looks like:

https://user-images.githubusercontent.com/15271942/164838809-22922935-df1e-4bf0-9361-d9a4f66dea34.mp4

**Comment by glwagner** [comment]
Here's a visualization of the initial condition:

https://user-images.githubusercontent.com/15271942/164845058-d7c14774-89db-4eba-957f-7d87204a5ccb.mp4

**Comment by glwagner** [comment]
This is another way to see the boundary conditions with zonally-averaged time-latitude Hovmoller diagrams:

zonally_averaged_one_degree_boundary_conditions

**Comment by glwagner** [comment]
It looks like there is a problem with GM. Taking 1 second time-steps without GM, `idealized_one_degree_simulation.jl` produces 

```julia
julia> include("near_global_one_degree.jl")
underlying_grid = LatitudeLongitudeGrid(arch; size = (Nx, Ny, Nz), halo = (4, 4, 4), latitude, z, longitude = (-180, 180), precompute_metrics = true) = 360×150×48 LatitudeLongitudeGrid{Float64, Periodic, Bounded, Bounded} on CPU with 4×4×4 halo and with precomputed metrics
├── longitude: Periodic λ ∈ [-180.0, 180.0) regularly spaced with Δλ=1.0
├── latitude:  Bounded  φ ∈ [-75.0, 75.0]   regularly spaced with Δφ=1.0
└── z:         Bounded  z ∈ [-5244.5, 0.0]  variably spaced with min(Δz)=10.0, max(Δz)=410.5
┌ Warning: WENO on a curvilinear stretched coordinate is not validated, use at your own risk!!
└ @ Oceananigans.Advection ~/Projects/dev/Oceananigans.jl/src/Advection/weno_fifth_order.jl:160
[ Info: Initializing simulation...
┌ Info: Iteration: 0, time: 0 seconds, wall time: 5.177 seconds
│ ├── max(u): (0.00e+00, 0.00e+00, 0.00e+00) m s⁻¹
│ ├── extrema(T): (0.31, 30.00) ᵒC
└ └── max|η|: 0.00e+00 m
[ Info:     ... simulation initialization complete (388.072 ms)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (28.553 seconds).
┌ Info: Iteration: 1, time: 1 second, wall time: 28.596 seconds
│ ├── max(u): (1.94e-10, 4.57e-05, 1.03e-06) m s⁻¹
│ ├── extrema(T): (0.31, 30.00) ᵒC
└ └── max|η|: 1.03e-06 m
┌ Info: Iteration: 2, time: 2 seconds, wall time: 1.632 seconds
│ ├── max(u): (9.46e-09, 9.13e-05, 2.06e-06) m s⁻¹
│ ├── extrema(T): (0.31, 30.00) ᵒC
└ └── max|η|: 3.09e-06 m
[ Info: Simulation is stopping. Model iteration 3 has hit or exceeded simulation stop iteration 3.
┌ Info: Iteration: 3, time: 3 seconds, wall time: 843.428 ms
│ ├── max(u): (2.48e-08, 1.37e-04, 3.09e-06) m s⁻¹
│ ├── extrema(T): (0.31, 30.00) ᵒC
└ └── max|η|: 6.17e-06 m
```

When `IsopycnalSkewSymmetricDiffusivity` is included, however, we find

```julia
julia> include("near_global_one_degree.jl")
underlying_grid = LatitudeLongitudeGrid(arch; size = (Nx, Ny, Nz), halo = (4, 4, 4), latitude, z, longitude = (-180, 180), precompute_metrics = true) = 360×150×48 LatitudeLongitudeGrid{Float64, Periodic, Bounded, Bounded} on CPU with 4×4×4 halo and with precomputed metrics
├── longitude: Periodic λ ∈ [-180.0, 180.0) regularly spaced with Δλ=1.0
├── latitude:  Bounded  φ ∈ [-75.0, 75.0]   regularly spaced with Δφ=1.0
└── z:         Bounded  z ∈ [-5244.5, 0.0]  variably spaced with min(Δz)=10.0, max(Δz)=410.5
┌ Warning: WENO on a curvilinear stretched coordinate is not validated, use at your own risk!!
└ @ Oceananigans.Advection ~/Projects/dev/Oceananigans.jl/src/Advection/weno_fifth_order.jl:160
[ Info: Initializing simulation...
┌ Info: Iteration: 0, time: 0 seconds, wall time: 259.679 ms
│ ├── max(u): (0.00e+00, 0.00e+00, 0.00e+00) m s⁻¹
│ ├── extrema(T): (0.31, 30.00) ᵒC
└ └── max|η|: 0.00e+00 m
[ Info:     ... simulation initialization complete (319.597 ms)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (1.116 seconds).
┌ Info: Iteration: 1, time: 1 second, wall time: 1.121 seconds
│ ├── max(u): (1.94e-10, 4.57e-05, 1.03e-06) m s⁻¹
│ ├── extrema(T): (0.31, 30.00) ᵒC
└ └── max|η|: 1.03e-06 m
┌ Info: Iteration: 2, time: 2 seconds, wall time: 998.405 ms
│ ├── max(u): (9.46e-09, 9.13e-05, 2.06e-06) m s⁻¹
│ ├── extrema(T): (-16387479425.34, 18491107419.68) ᵒC
└ └── max|η|: 3.09e-06 m
[ Info: Simulation is stopping. Model iteration 3 has hit or exceeded simulation stop iteration 3.
┌ Info: Iteration: 3, time: 3 seconds, wall time: 997.956 ms
│ ├── max(u): (9.82e+04, 7.85e+04, 2.49e+03) m s⁻¹
│ ├── extrema(T): (-1654340357512160890046119936.00, 2003185335144995241362194432.00) ᵒC
└ └── max|η|: 2.34e+03 m
```

so the temperature is going out of control. This precedes the velocity issue so the problem is specifically with temperature.

**Comment by glwagner** [comment]
During the coefficients down to 1e-9 (!!) allows 3 stable time steps _when continents / bathymetry are removed_:

```julia
julia> include("idealized_one_degree_simulation.jl")
grid = 360×150×48 LatitudeLongitudeGrid{Float64, Periodic, Bounded, Bounded} on CPU with 4×4×4 halo and with precomputed metrics
├── longitude: Periodic λ ∈ [-180.0, 180.0) regularly spaced with Δλ=1.0
├── latitude:  Bounded  φ ∈ [-75.0, 75.0]   regularly spaced with Δφ=1.0
└── z:         Bounded  z ∈ [-5244.5, 0.0]  variably spaced with min(Δz)=10.0, max(Δz)=410.5
┌ Warning: WENO on a curvilinear stretched coordinate is not validated, use at your own risk!!
└ @ Oceananigans.Advection ~/Projects/dev/Oceananigans.jl/src/Advection/weno_fifth_order.jl:160
[ Info: Initializing simulation...
┌ Info: Iteration: 0, time: 0 seconds, wall time: 226.942 ms
│ ├── max(u): (0.00e+00, 0.00e+00, 0.00e+00) m s⁻¹
│ ├── extrema(T): (0.31, 30.00) ᵒC
└ └── max|η|: 0.00e+00 m
[ Info:     ... simulation initialization complete (288.084 ms)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (1.446 seconds).
┌ Info: Iteration: 1, time: 20 minutes, wall time: 1.452 seconds
│ ├── max(u): (1.32e-16, 5.40e-02, 6.44e-04) m s⁻¹
│ ├── extrema(T): (0.31, 30.00) ᵒC
└ └── max|η|: 5.93e-01 m
┌ Info: Iteration: 2, time: 40 minutes, wall time: 945.169 ms
│ ├── max(u): (1.39e-02, 1.06e-01, 1.05e-03) m s⁻¹
│ ├── extrema(T): (0.05, 30.00) ᵒC
└ └── max|η|: 1.33e+00 m
[ Info: Simulation is stopping. Model iteration 3 has hit or exceeded simulation stop iteration 3.
┌ Info: Iteration: 3, time: 1 hour, wall time: 945.031 ms
│ ├── max(u): (3.55e-02, 1.52e-01, 1.37e-03) m s⁻¹
│ ├── extrema(T): (0.15, 30.00) ᵒC
└ └── max|η|: 2.08e+00 m
```

But when they're put back, it still blows up (even with coefficients 1e-9):

```julia
julia> include("idealized_one_degree_simulation.jl")
grid = ImmersedBoundaryGrid on:
    architecture: CPU()
            grid: 360×150×48 LatitudeLongitudeGrid{Float64, Periodic, Bounded, Bounded} on CPU with 4×4×4 halo and with precomputed metrics
   with immersed: GridFittedBottom{OffsetArrays.OffsetMatrix{Float64, Matrix{Float64}}}
┌ Warning: WENO on a curvilinear stretched coordinate is not validated, use at your own risk!!
└ @ Oceananigans.Advection ~/Projects/dev/Oceananigans.jl/src/Advection/weno_fifth_order.jl:160
[ Info: Initializing simulation...
┌ Info: Iteration: 0, time: 0 seconds, wall time: 256.442 ms
│ ├── max(u): (0.00e+00, 0.00e+00, 0.00e+00) m s⁻¹
│ ├── extrema(T): (0.31, 30.00) ᵒC
└ └── max|η|: 0.00e+00 m
[ Info:     ... simulation initialization complete (317.156 ms)
[ Info: Executing initial time step...
[ Info:     ... initial time step complete (1.252 seconds).
┌ Info: Iteration: 1, time: 20 minutes, wall time: 1.258 seconds
│ ├── max(u): (7.36e-02, 5.18e-02, 7.93e-04) m s⁻¹
│ ├── extrema(T): (0.31, 30.00) ᵒC
└ └── max|η|: 3.73e-01 m
┌ Info: Iteration: 2, time: 40 minutes, wall time: 1.003 seconds
│ ├── max(u): (2.83e-01, 1.13e-01, 1.65e-03) m s⁻¹
│ ├── extrema(T): (-6.76, 50.82) ᵒC
└ └── max|η|: 8.05e-01 m
[ Info: Simulation is stopping. Model iteration 3 has hit or exceeded simulation stop iteration 3.
┌ Info: Iteration: 3, time: 1 hour, wall time: 993.451 ms
│ ├── max(u): (6.23e-01, 2.08e-01, 2.96e-03) m s⁻¹
│ ├── extrema(T): (-4.27, 50.74) ᵒC
└ └── max|η|: 1.22e+00 m
```

so there is almost _certainly_ a bug with how `IsopycnalSkewSymmetricDiffusivity` is implemented.

@navidcy @sandreza 

Now I know we have tested the diffusivity a fair amount in `y, z` domains, but have we tested `x, z`? I don't think we've ever tried 3D either. We should probably put GM in a double gyre and see what happens (with and without immersed boundaries).

**Comment by glwagner** [comment]
I added a script for looking at coarse, 3D baroclinic adjustment with GM starting with either a gradient in `y` or in `x`: https://github.com/CliMA/Oceananigans.jl/blob/0d04d4014c9ecc06deb158c42a2b3e3cd7699b31/validation/mesoscale_turbulence/coarse_baroclinic_adjustment.jl

It was tricky to get this to work in the first place; it seems we need diffusion to stabilize GM right now. I settled on adding biharmonic diffusivity with a time scale of 10 days.

The script computes a mean in the direction perpendicular to the initial gradient.

When the gradient is initially in `y`:

https://user-images.githubusercontent.com/15271942/164878441-a0660d6e-3241-4ec4-a473-cf0c65c70fa6.mp4

When the gradient is in `x`:

https://user-images.githubusercontent.com/15271942/164878658-18604606-9374-4008-894e-29ee7e49782a.mp4

From this it seems that the diffusivity is at least doing the right thing (but the stencil may still be non-optimal).

We also need to test on a lat-lon grid.

**Comment by glwagner** [comment]
Now there's `coarse_lat_lon_baroclinic_adjustment.jl` as well.

This time `x` is not quite the same as `y` because grid distances change with latitude for a regular lat-lon grid. It's a little sketchier; heres the case with an y-gradient (meridional gradient):

https://user-images.githubusercontent.com/15271942/164879429-f513203b-8ac8-48d6-94fd-d2e25d794e69.mp4

and an x gradient (longitudinal gradient):

https://user-images.githubusercontent.com/15271942/164879433-e0b88ef0-d1aa-43dd-b6d8-9c758780887b.mp4

It seems the biharmonic diffusion is crucial.

**Comment by francispoulin** [comment]
> I added a script for looking at coarse, 3D baroclinic adjustment with GM starting with either a gradient in `y` or in `x`: https://github.com/CliMA/Oceananigans.jl/blob/0d04d4014c9ecc06deb158c42a2b3e3cd7699b31/validation/mesoscale_turbulence/coarse_baroclinic_adjustment.jl
> 
> It was tricky to get this to work in the first place; it seems we need diffusion to stabilize GM right now. I settled on adding biharmonic diffusivity with a time scale of 10 days.
> 
> The script computes a mean in the direction perpendicular to the initial gradient.
> 
> When the gradient is initially in `y`:
> 
>  coarse_baroclinic_adjustment_y.mp4 
> When the gradient is in `x`:
> 
>  coarse_baroclinic_adjustment_x.mp4 
> From this it seems that the diffusivity is at least doing the right thing (but the stencil may still be non-optimal).
> 
> We also need to test on a lat-lon grid.

I see that the two animations are similar, but not exactly the same.  I presume the initial condigurations aren't simply rotated by 90 degrees?

**Comment by glwagner** [comment]
Bah sorry @francispoulin I meant to respond to your comment, but accidentally edited instead. Then I tried to delete my edit, but accidentally deleted the original comment!

**Comment by glwagner** [comment]
It needs relatively strong biharmonic viscosity / diffusivity to stabilize the GM.

**Comment by francispoulin** [comment]
Thanks @glwagner .  It seems that the `ramp` changes with the direction of the gradient, but the tracer doesn't.  That's why we have a difference.  Thanks.

This is a rather complicated problem that you are investigating and I wonder whether getting a 3D test for `IsopycnalSkewSymmetricDiffusivity` might be an easier way to find the bug?  Just a thought.

**Comment by glwagner** [comment]
> Thanks @glwagner . It seems that the `ramp` changes with the direction of the gradient, but the tracer doesn't. That's why we have a difference. Thanks.
> 
> This is a rather complicated problem that you are investigating and I wonder whether getting a 3D test for `IsopycnalSkewSymmetricDiffusivity` might be an easier way to find the bug? Just a thought.

Agree! Partly I'm not sure what to expect. It seems that GM is noisy in general, so I'm not sure whether our implementation is "more noisy" than it should be.

These examples are 3D but maybe you mean with an x, y gradient rather than in x or y independently? I was thinking that maybe a 3D double gyre with GM might be helpful, what do you think?

**Comment by francispoulin** [comment]
Sorry,  you mentioned the tests only tested `y,z`, but not `x,z` or `x,y,z`.  I thought if those tests could be done that might help to identify the problem.  Maybe?

A double gyre might help but that introduces boundaries, or at least more boundaries, so maybe more compliated?

It is interesting that we get an instability in a lat-lon grid but not a Cartesian grid.  If you re-scale the lengths down say by a factor of 10, or even 100, these two should agree.  I wonder if we still get an instability?  If we did in the lat-lon grid but not the Cartesian one, maybe that would point to a problem with diffusing on a sphere?  Just a thought.

**Comment by glwagner** [comment]
@christophernhill pointed me to this MITgcm setup: https://github.com/MITgcm/MITgcm/tree/c6adb04bff0a8860ffa46b74e8a11a6a6228274f/verification/front_relax/input

which is very similar and has no horizontal diffusion. However, it does look like it might have horizontal biharmonic viscosity with [`viscA4=1e11`](https://github.com/MITgcm/MITgcm/blob/c6adb04bff0a8860ffa46b74e8a11a6a6228274f/verification/front_relax/input/data#L9) --- can someone else confirm that I'm interpreting that correctly?

I'll re-run these tests with just biharmonic viscosity but no horizontal diffusion (laplacian or biharmonic) and see what happens. Maybe if the results are noisy then we can be sure there's an issue...

**Comment by glwagner** [comment]
Two more runs with `κ_symmetric=0.0`: https://github.com/CliMA/Oceananigans.jl/pull/2463/commits/6cf05a19071071646ea81db6dba7f9741530ff35

## x-gradient / longitudinal gradient:

https://user-images.githubusercontent.com/15271942/165175271-1f442db9-2419-4163-b51b-49bc2636d2e9.mp4

## y-gradient / latitudinal gradient:

https://user-images.githubusercontent.com/15271942/165175350-50072c66-7c1c-475a-935a-3ef7b17c4ccc.mp4

Things are still a little sketchy.

These were run because @sandreza reports that MITgcm creates the vertical symmetric diffusivity term implicitly.

**Comment by sandreza** [comment]
@xiaozhour and I looked into some direct comparisons with MITgcm and found a few differences that we are continuing to investigate. One thing worth noting is that in the MITgcm computation the xz and yz components of the diffusivity tensor that act on vertical gradients are directly set to zero, whereas the zz component is treated implicitly.

In the simulation below I increase the initial horizontal gradient by a factor of two and set the initial tracer to be constant. A numerical instability develops eventually leading to NaNs in the tracer field. By setting the Redi-component to zero the solution remains stable. The file is located here: https://github.com/CliMA/Oceananigans.jl/blob/glw-as/near-global-one-degree/validation/mesoscale_turbulence/modified_baroclinic_adjustment.jl

https://user-images.githubusercontent.com/22668662/165390858-211bce67-fb10-4ae4-8e60-3516b49e5031.mp4

**Comment by navidcy** [comment]
I have seen this sort of behaviour before... :(
We fixed it at some point. But I guess it came back to haunt us...

**Comment by glwagner** [comment]
Is the buoyancy field also so noisy in MITgcm? I'll work on treating the dz k dz c term from the Redi diffusivity implicitly. But it seems that there might still be a problem.

@navidcy did our changes get reverted somehow?

**Comment by glwagner** [comment]
Do you have animations from the MITgcm run @sandreza ?

**Comment by xiaozhour** [comment]
https://user-images.githubusercontent.com/13021515/165544450-152c1874-0eac-4ea6-809a-e84420ec5fc2.mp4

@glwagner

**Comment by glwagner** [comment]
Nice!!

Two immediate observations:

1. The MITgcm solution is noisy too
2. They are _very different_ at the top and bottom boundary. Maybe our issue is boundary interaction.

**Comment by glwagner** [comment]
@xiaozhour what resolution is this? Looks relatively high.

**Comment by glwagner** [comment]
> It uses the same resolution and parameters as in Andre's run.

Ok, and just for documentation that domain is

```juila
# Domain
Lx = 1000kilometers
Ly = 1000kilometers  # north-south extent [m]
Lz = 1kilometers     # depth [m]
Nx = 20
Ny = 20
Nz = 20
save_fields_interval = 0.5day
stop_time = 30days
Δt = 20minutes

zfaces = -Lz .+ (-0.5 * (cos.(range(0.0, π, length=Nz + 1))) .+ 0.5) .* Lz # range(-Lz, 0.0, length = Nz+1)
zfaces = range(-Lz, 0.0, length=Nz + 1)

grid = RectilinearGrid(architecture;
    topology=(Bounded, Bounded, Bounded),
    size=(Nx, Ny, Nz),
    x=(-Lx / 2, Lx / 2),
    y=(-Ly / 2, Ly / 2),
    z=zfaces,
    halo=(3, 3, 3))
```

**Comment by xiaozhour** [comment]
This is the advective form of GM/Redi in MITgcm..

https://user-images.githubusercontent.com/13021515/165631167-8e539435-e3c4-4b2f-8dbd-cf3ea90329d5.mp4

**Comment by glwagner** [comment]
One important difference could be that we don't have the "`GM_small_number` parameter that MITgcm has:

https://github.com/MITgcm/MITgcm/blob/9bcb4977016b9cc234328aa357b105d227417e43/pkg/gmredi/gmredi_slope_psi.F#L296

**Comment by francispoulin** [comment]
> One important difference could be that we don't have the "`GM_small_number`" parameter that MITgcm has:
> 
> https://github.com/MITgcm/MITgcm/blob/9bcb4977016b9cc234328aa357b105d227417e43/pkg/gmredi/gmredi_slope_psi.F#L296

And whawt does that parameter do?

**Comment by glwagner** [comment]
> > One important difference could be that we don't have the "`GM_small_number`" parameter that MITgcm has:
> > https://github.com/MITgcm/MITgcm/blob/9bcb4977016b9cc234328aa357b105d227417e43/pkg/gmredi/gmredi_slope_psi.F#L296
> 
> And whawt does that parameter do?

Hmm, I guess it prevents that taper field from becoming too large? Honestly, I'm not 100% sure. You might read fortran better than me 😆

**Comment by glwagner** [comment]
The problem with overturns appearing near the surface may have to do with how tapering interacts with the boundary. Since bz = 0 on the boundary, and Sx = bx / bz, the slope is larger by construction in the first grid cell. For the problem we are looking at, it appears that tapering does indeed get activated at the surface. This means that the flux is proportionally smaller in the first grid point below the surface, meaning that the isopycnal tilt faster in the interior than at the surface, leading to overturns. This was discovered with @navidcy

**Comment by sandreza** [comment]
It does seem like the additional fix for the boundary condition on GM is essential. I recently tried the 1 degree setup (switching to v0.75.3 of Oceananigans) and saw that there are still stability issues. Given that the solution blows up within 10 iterations it's probably more worthwhile to use the CPU for debugging purposes. It took 50 minutes to precompile on the GPU with TEOS10 + GM + Biharmonic viscosity + Vertical Diffusivity + Convective Adjustment

**Comment by sandreza** [comment]
@simone-silvestri identified (at least part) of the reason why it took so long to compile the solution: The free-surface solver was spending too much time trying to construct the approximate sparse inverse. This is not the whole story, since it does not explain why adding additional closures increased the start up time; however, removing the nice preconditioner is worse for taking many timesteps

**Comment by simone-silvestri** [comment]
We maybe want to optimize the construction of the `SparseInverse`.
I naively wrote a serial code to do that because I didn't know that `CUDA.jl` had implemented a GPU version of the `qr` factorization. Since the construction of the preconditioner is embarrassingly parallelizable we can shift it to the GPU easily

**Comment by glwagner** [comment]
it looks like `validation/near_global_lat_lon/near_global_quarter_degree.jl` was updated elsewhere so we need to resolve merge conflicts

**Comment by simone-silvestri** [comment]
When do we plan to merge this? If we merge it I can use the same methods for the 1/12

**Comment by glwagner** [comment]
I think we can merge now. Maybe it'd be nice to have the realistic case in a similar file as `one_degree_setups/idealized_near_global_one_degree.jl` (if it isn't there already)?

https://github.com/CliMA/Oceananigans.jl/blob/glw-as/near-global-one-degree/validation/near_global_lat_lon/one_degree_setups/idealized_near_global_one_degree.jl

**Comment by simone-silvestri** [comment]
I ll fix the conflicts and merge ok?

**Comment by glwagner** [comment]
sounds good

**Review by simone-silvestri — APPROVED** [review]

**Pull request merged**